Add post about variable storage optimizations (#177)
boa-dev/boa-dev.github.io: blog/2024-10-01-local-variables.md

---
layout: post
tags: [post, internals]
title: "How ECMAScript Engines Optimize Your Variables"
authors: boa-dev
---

In this post we dive into how ECMAScript engines store variables.

> We go over storage optimizations and learn about scope analysis.

Reviewer: ```suggestion
In this post, we will dive into how ECMAScript engines store variables, 
go over storage optimizations, and learn about scope analysis.
```

> If you are an ECMAScript developer you will get some practical tips to improve the performance of your code.

Reviewer: ```suggestion
If you are an ECMAScript developer, you will get some practical tips to improve the performance of your code.
```

> If you write your own ECMAScript engine or any interpreter or compiler you might get some implementation ideas.

Reviewer: ```suggestion
If you write your own ECMAScript engine or any interpreter/compiler, you might get some implementation ideas.
```

<!--truncate-->

Before we start, let's get some disclaimers out of the way:

- This post is written about optimizations in the Boa ECMAScript engine.
  Other engines might do things different, so take everything with a grain of salt.
- This post omits some implementation details to focus on the most relevant parts.
- This post contains data structures written in pseudo Rust.

>   These are only for visualization, and you do not need to understand Rust.

Reviewer: ```suggestion
- This post contains data structures written in pseudo Rust
  that are only for visualization, so you should not need to understand Rust.
```

## Scopes and Variables

To start us off, let's refresh our understanding of variables and scopes in ECMAScript.
Variables are identifiers in our code that we can use to store and retrieve values.
Scopes describe the areas of code in which variables are visible.
You might associate scopes with curly braces.

Let's look at an example to visualize scopes:

```js

> const a = 1;

Reviewer: Would it make sense to also add global scope indicators like block scope, just to be explicit.

Author: I though about it, but I left that out intentionally. I think it would overload the examples a bit and distract too much from the main point.

console.log(a); // 1

{ // <- start of a block scope
    const a = 2;
    console.log(a); // 2
} // <- end of a block scope
```

We declare and initialize two variables with the identifier `a`.
Even though both have the same identifier, they are different variables.
The variable with the value `1` is declared in the global scope.
The variable with the value `2` is declared in a block scope.

> In this example we always find the variable that we want to access in the current scope.

Reviewer: ```suggestion
In this example, we always find the variable that we want to access in the current scope.
```

But what when the variable we access in not declared in the current scope?

Let's modify our example to see what happens in that case:

```js
const a = 1;

{
    const b = 2;
    console.log(a); // 1
    console.log(b); // 2
}
```

> In this example our two variables have different identifiers.

Reviewer: ```suggestion
In this example, our two variables have different identifiers.
```

Notice that when we access the variable `a` from the block scope, its value is resolved as expected.
This is because scopes are nested.
When we cannot find a variable in the current scope, we look for the same identifier in the outer scope.

> In this case we have to look for `a` in the block scope and then in the global scope.

Reviewer: ```suggestion
In this case, we have to look for `a` in the block scope and then in the global scope.
```

Let's look at a more complex example:

```js
const a = 1;
console.log(a); // 1

function f() { // <- start of a function scope
    var a = 2;
    console.log(a); // 2

    { // <- start of a block scope
        let a = 3;
        console.log(a); // 3
    } // <- end of a block scope

    console.log(a); // 2
} // <- end of a function scope

f();

console.log(a); // 1
```

You can see that variables are tied to their scopes.
All three variables `a` never change their values.

> They just exist in their respective scopes and as soon as the scope has ended they are no longer accessible.

Reviewer: ```suggestion
The variables just exist in their respective scopes, and as soon as the scope has ended, they are no longer accessible.
```

struct Scope {
    index: u32,
    variables: HashMap<Identifier, Variable>,
    outer: Scope,
}

struct Variable {
    index: u32,
}
```

While this data structure still works based on self-referential hashmaps, we only need it before running code.
Instead of doing a lookup on every variable access at runtime, we just have to do it once.

## Local Variables

Our previous optimization changed the data structure for our variables and scopes.
But we can come up with further optimizations if we look at the usage of variables in the code.

Let's take a look at this example:

```js
function addOne(a) {
    const one = 1;
    return one + a;
}
addOne(2);
```

Currently, we store `a` and `one` in our scopes and access them when performing the addition.
But why do we need the special data structure for variables and scopes at all?
What if we could just store the variables directly where we need them?

A typical ECMAScript engine uses a virtual machine (VM) to execute your code.
VMs use dedicated memory for values they operate on; a stack or registers.
For the purpose of this post, we use registers, but the stack would work in the same way.
Let's try to use registers to store variables.

> While compiling the ECMAScript code into operations for our VM we assign each variable to a register.

Reviewer: ```suggestion
While compiling the ECMAScript code into operations for our VM, we assign each variable to a register.
```

## Scope Analysis

### Finding Variables accessed by other Functions

To determine which variables can be stored in registers, we analyze them prior to code execution.
We can reuse our previously established scope structure based on hashmaps.
It just needs some additional information to make our analysis work.
Each scope needs to be flagged to indicate if it is a function scope.
This is important, because we have to track if a variable is ever accessed from a nested function.
Additionally, each variable needs a flag to indicate if it can be a local variable.

Our adjusted scope structure looks like this:

```rust
struct Scope {
    index: u32,
    function: bool, // <- This field is new
    variables: HashMap<Identifier, Variable>,
    outer: Scope,
}

struct Variable {
    index: u32,
    local: bool, // <- This field is new
}
```

We set the `function` flag on each scope when it is created.
The `local` flag on the variables is `true` by default.
After creating all the scopes and filling them with their variables, we traverse the ECMAScript code.
Every time we find a variable access, we check which variable in which scope is being accessed.

> If a variable is not in the current scope we go to the `outer` scope.

Reviewer: ```suggestion
If a variable is not in the current scope, we go to the `outer` scope.
```

If any of the scopes we see until we find the variable is a function scope, we set the `local` flag to `false`.

Let's visualize the scope analysis by writing out the scopes for this example:

```js
function addOneBuilder() {
    const one = 1;
    return (a) => {
        return one + a;
    };
}
```

These are our scopes before the scope analysis.
Notice that we start from the scope of the arrow function, since we work our way out from the most nested scope:

```rust
Scope {
    function: true
    variables: [
        "a": {
            index: 0,
            local: true,
        }
    ]
    outer: Scope {
        function: true
        variables: [
            "one": {
                index: 0,
                local: true,
            }
        ]
    }
}
```

Now we apply our scope analysis.
During the access to variable `one` in the arrow function, we pass the function scope of the arrow function.

> This indicates, that this variable cannot be local:

Reviewer: ```suggestion
This indicates that this variable cannot be local:
```

```rust
Scope {
    function: true
    variables: [
        "a": {
            index: 0,
            local: true, // <- Still `true`
        }
    ]
    outer: Scope {
        function: true
        variables: [
            "one": {
                index: 0,
                local: false, // <- Set to `false`
            }
        ]
    }
}
```

> After the scope analysis is finished we compile our code for the VM.

Reviewer: ```suggestion
After the scope analysis is finished, we compile our code for the VM.
```

When we encounter a variable that can be local, we assign it a register.

> When we encounter a variable that cannot be local, we use or old VM operations for our scopes storage.

Reviewer: ```suggestion
When we encounter a variable that cannot be local, we use the old VM operations for our scopes storage.
```

    Direct calls to `eval` have access to the current variables.
    Since any code could be executed in `eval` we cannot do proper scope analysis on any variables in such cases.

    An example of direct `eval` usage would be this:

    ```js
    function f() {
        const a = 1;
        eval("function nested() {console.log(a)}; nested();");
    }
    f();
    ```

    Our solution is this case is to mark every variable in the scopes where the direct `eval` call is as non-local.

- Usage of the `with` statement.
    Variable identifiers inside a `with` statement are not static.
    A variable identifier could be the access to a variable, but it also could be the access to an object property.

    See this example:

    ```js
    function f() {
        const a1 = 1;
        for (let i = 0; i < 2; i++) {
            with ({ [`a${i}`]: 2 }) {
                console.log(a1);
            }
        }
    }
    f();
    ```

    In the first loop execution `a1` is the variable.
    In the second loop execution `a1` is the object property.
    As a result of this behavior, every variable accessed inside a `with` statement cannot be local.

## Conclusion

> After implementing local variables in Boa we saw significant performance improvements in our benchmarks.

Reviewer: ```suggestion
After implementing local variables in Boa, we saw significant performance improvements in our benchmarks.
```

Our overall benchmark scope improved by more than 25%.
In one specific benchmark the scope increased by over 70%.
Notice that Boa is not the most performant engine yet. 
There are probably other optimizations relating to variable storage that we have not implemented yet.

> You might have already picked up some practical tips to potentially improve to performance of your ECMAScript code.

Reviewer: ```suggestion
Hopefully, you might have already picked up some practical tips to potentially improve to performance of your ECMAScript code.
```
